stdlib/LinearAlgebra/src/symmetriceigen.jl

```diff
@@ -63,17 +94,42 @@ function eigen(A::RealHermSymComplexHerm, vl::Real, vh::Real)
     eigen!(eigencopy_oftype(A, S), vl, vh)
 end
 
-function eigvals!(A::RealHermSymComplexHerm{<:BlasReal,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing)
-    vals = LAPACK.syevr!('N', 'A', A.uplo, A.data, 0.0, 0.0, 0, 0, -1.0)[1]
+
+function eigvals!(A::RealHermSymComplexHerm{<:BlasReal,<:StridedMatrix}, lapack_method::Symbol=:syevr; sortby::Union{Function,Nothing}=nothing)
+    if lapack_method === :syevr
+      vals = LAPACK.syevr!('N', 'A', A.uplo, A.data, 0.0, 0.0, 0, 0, -1.0)[1]
+    elseif lapack_method === :syev
+      vals = LAPACK.syev!('N', A.uplo, A.data)
+    elseif lapack_method === :syevd
+      vals = LAPACK.syevd!('N', A.uplo, A.data)
+    else
```

Comment:
This creates type instability in `eigvals!` due to type instability of `LAPACK.syev!` and `LAPACK.syevd!`.

When computing both the eigenvalues and eigenvectors, all routines return a vector (eigenvalues) and a matrix (eigenvectors).

When computing only the eigenvalues, `syevr!` return the eigenvalues and a dummy matrix which makes it type stable.
https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/lapack.jl#L5263
But, `syev!` and `syevd! returns only the eigenvalues.
https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/lapack.jl#L5201

* Option 1: Change output of `LAPACK.syev!` and `syevd!` to return dummy array and make they type stable. (Is it breaking?)
* Option 2: Add type assertion
* Option 3: Don't create a `lapack_method` interface for `eigvals`, just keep `eigen`

What do you think? Are there any other solutions?